### PR TITLE
Support for call the flush callback at regular intervals using the  plugin read callback.

### PR DESCRIPTION
**Pull request by manuelluis**
It's the version of https://github.com/collectd/collectd/pull/1044 using the  plugin read callback.

### CODE REVIEW

**Comment by manuelluis** [comment]
TODO / debate:
- ``plugin_flush`` not check for errors in the callback or missing names.
- the use of malloc and free to create: "flush/name"

**Comment by octo** [comment]
@manuelluis Looks awesome, thank you! I'll give others a day or so to comment before merging though.

File: src/daemon/plugin.c [hunk 1]
```diff
@@ -1295,11 +1301,99 @@ int plugin_register_write (const char *name,
...
+		flush_name = (char *) malloc (sizeof (char) *
+			(name_size + prefix_size + 1));
+		if (flush_name == NULL)
+		{
+			ERROR ("plugin_register_flush: malloc failed.");
+			plugin_unregister (list_flush, name);
+			return (-1);
+		}
+
+		sstrncpy (flush_name, flush_prefix, prefix_size + 1);
```
Comment (octo) [inline src/daemon/plugin.c]: These 12 or so lines constructing `flush_name` appear again in `plugin_unregister_flush()`. How about a function like this:?

```c
static char *plugin_flush_callback_name (char const *base_name);
```

I see the following benefits:

* Less code duplication.
* Makes it harder to break by partial code changes.
* `if { … }` blocks become shorter.
* `ERROR()` reporting when `malloc()` fails can be done in the function.

I'm not super happy with the above name, as "callback" is not a very good name for what I try to describe, but I can't think of a better alternative atm.

File: src/daemon/plugin.c [hunk 2]
```diff
@@ -1295,11 +1301,99 @@ int plugin_register_write (const char *name,
...
+		sstrncpy (flush_name + prefix_size, name, name_size + 1);
+
+		cb = (flush_callback_t *)malloc(sizeof(flush_callback_t));
```
Comment (octo) [inline src/daemon/plugin.c]: Nit: the preferred style of `{m,re}alloc()` calls is:

```c
cb = malloc (sizeof (*cb));
```

I.e. remove all type information and let the compiler infer it. This makes refactorings easier and the line less cluttered.

File: src/daemon/plugin.c [hunk 3]
```diff
@@ -1295,11 +1301,99 @@ int plugin_register_write (const char *name,
...
+			/* callback  = */ plugin_flush_timeout_callback,
+			/* interval  = */ ctx.flush_interval,
+			/* user data = */ &ud);
+
+		sfree(flush_name);
+		if (status != 0)
+		{
+			sfree(cb->name);
+			sfree(cb);
+			plugin_unregister (list_flush, name);
```
Comment (octo) [inline src/daemon/plugin.c]: Not sure if that's the most intuitive failure mode. If the installing the regular flush fails, letting the flush callback itself register seems more intuitive IMHO.

**Comment by pyr** [comment]
Hi @manuelluis. Thanks for this, this will be of great help for `write_riemann` :-) :+1:

**Comment by octo** [comment]
Thanks @manuelluis

**Pull request merged**